src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs

```diff
@@ -1132,10 +1132,23 @@ private BoundExpression BindSizeOf(SizeOfExpressionSyntax node, DiagnosticBag di
 
             bool typeHasErrors = type.IsErrorType();
 
-            if (!typeHasErrors && type.IsManagedType)
+            if (!typeHasErrors)
             {
-                diagnostics.Add(ErrorCode.ERR_ManagedAddr, node.Location, type);
-                typeHasErrors = true;
+                if (type.IsManagedType)
+                {
+                    diagnostics.Add(ErrorCode.ERR_ManagedAddr, typeSyntax.Location, type);
+                    typeHasErrors = true;
+                }
+                else if (type.GetArity() != 0)
+                {
+                    var unsupported = Compilation.LanguageVersion < MessageID.IDS_FeatureUnmanagedGenericStructs.RequiredVersion();
+                    if (unsupported)
+                    {
+                        // PROTOTYPE
+                        MessageID.IDS_FeatureUnmanagedGenericStructs.CheckFeatureAvailability(Compilation.LanguageVersion, diagnostics, typeSyntax.Location);
+                        typeHasErrors = true;
```

Comment:
Wondering if perhaps it's fine to give the language version error as well as the additional errors that are supposed to be gated here by `typeHasErrors` and similar flags.

That is, just replace the whole block containing `var unsupported = ...` with just the call to `CheckFeatureAvailability`.

Reply:
Take a look at the other existing usages of `CheckFeatureAvailability`. A good example is here:

[https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L1904](https://github.com/dotnet/roslyn/blob/master/src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs#L1904)

The pattern is to just call the function and let it add an entry to the `DiagnosticBag`. The returned type / expression is not otherwise modified. That pattern is used in a number of places. I would follow it here too. 

---
In reply to: [242373027](https://github.com/dotnet/roslyn/pull/31602#discussion_r242373027) [](ancestors = 242373027)

Reply:
Sure, it sounds like I'll just avoid setting `typeHasErrors`-like flags in response to feature unavailability.